File: TravelChain.java
import java.util.*;

public class TravelChain {

    public static String findLongestChainDestination(List<Pair<String, String>> itinerary) {
        if (itinerary == null || itinerary.isEmpty()) {
            return "";
        }

        if (itinerary.size() == 1 && itinerary.get(0).first.equals(itinerary.get(0).second)) {
            return "обратитесь к специалисту";
        }

        // Построение карты маршрутов и множества всех городов
        Map<String, String> routes = new HashMap<>();
        Set<String> cities = new HashSet<>();

        for (Pair<String, String> leg : itinerary) {
            routes.put(leg.first, leg.second);
            cities.add(leg.first);
            cities.add(leg.second);
        }

        // Поиск начального города (города, который является началом, но не концом ни одного другого маршрута)
        String startCity = null;
        for (String city : cities) {
            boolean isStart = true;
            for (Pair<String, String> leg : itinerary) {
                if (leg.second.equals(city)) {
                    isStart = false;
                    break;
                }
            }
            if (isStart) {
                startCity = city;
                break;
            }
        }

        // Если начальный город не найден, возвращаем пустую строку
        if (startCity == null) {
            return "";
        }

        // Проходим по маршруту от начального города до конечного
        String currentCity = startCity;
        String finalDestination = null;
        while (routes.containsKey(currentCity)) {
            finalDestination = routes.get(currentCity);
            currentCity = routes.get(currentCity);
        }

        return finalDestination;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Введите количество пар пунктов пребывания: ");
        int numPairs = scanner.nextInt();
        scanner.nextLine(); // Consume newline

        List<Pair<String, String>> itinerary = new ArrayList<>();
        for (int i = 0; i < numPairs; ++i) {
            System.out.print("Введите пару " + (i + 1) + " (Начальный пункт,Конечный пункт): ");
            String line = scanner.nextLine();
            String[] parts = line.split(",");

            if (parts.length == 2) {
                String start = parts[0].trim();
                String end = parts[1].trim();
                itinerary.add(new Pair<>(start, end));
            } else {
                System.out.println("Некорректный ввод. Пожалуйста, введите пару в формате 'Начальный пункт,Конечный пункт'.");
                i--; // Repeat input for this pair
            }
        }

        String result = findLongestChainDestination(itinerary);

        if (result.equals("обратитесь к специалисту")) {
            System.out.println(result);
        } else if (result.isEmpty()) {
            System.out.println("Маршрут не найден.");
        } else if (itinerary.size() == 1 && itinerary.get(0).first.equals(itinerary.get(0).second)) {
            System.out.println("обратитесь к специалисту");
        }
        else {
            System.out.println("Итоговый конечный пункт: " + result);
        }

        scanner.close();
    }

    // Helper class to represent a pair
    static class Pair<K, V> {
        public K first;
        public V second;

        public Pair(K first, V second) {
            this.first = first;
            this.second = second;
        }
    }
}

File: travel_chain.ts
import * as readline from 'readline';

function findLongestChainDestination(itinerary: [string, string][]): string {
    if (!itinerary || itinerary.length === 0) {
        return "";
    }

    if (itinerary.length === 1 && itinerary[0][0] === itinerary[0][1]) {
        return "обратитесь к специалисту";
    }

    // Построение карты маршрутов и множества всех городов
    const routes: { [key: string]: string } = {};
    const cities: Set<string> = new Set();
    for (const leg of itinerary) {
        routes[leg[0]] = leg[1];
        cities.add(leg[0]);
        cities.add(leg[1]);
    }

    // Поиск начального города (города, который является началом, но не концом ни одного другого маршрута)
    let startCity: string | null = null;
    cities.forEach(city => { // Изменен цикл for...of на forEach
        let isStart = true;
        for (const leg of itinerary) {
            if (leg[1] === city) {
                isStart = false;
                break;
            }
        }
        if (isStart) {
            startCity = city;
            return; // Замена break в цикле forEach
        }
    });

    // Если начальный город не найден, возвращаем пустую строку
    if (!startCity) {
        return "";
    }

    // Проходим по маршруту от начального города до конечного
    let currentCity: string | null = startCity;
    let finalDestination: string | null = null;
    while (currentCity && routes[currentCity]) {
        finalDestination = routes[currentCity];
        currentCity = routes[currentCity];
    }

    return finalDestination || "";
}

async function main() {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
    });

    const question = (query: string): Promise<string> => {
        return new Promise((resolve) => {
            rl.question(query, resolve);
        });
    };


    try {
        const numPairsStr = await question("Введите количество пар пунктов пребывания: ");
        const numPairs = parseInt(numPairsStr);

        if (isNaN(numPairs) || numPairs < 0) {
            console.log("Некорректный ввод количества пар.");
            return;
        }

        const itinerary: [string, string][] = [];
        for (let i = 0; i < numPairs; ++i) {
            const line = await question(`Введите пару ${i + 1} (Начальный пункт,Конечный пункт): `);
            const parts = line.split(",");

            if (parts.length === 2) {
                const start = parts[0].trim();
                const end = parts[1].trim();
                itinerary.push([start, end]);
            } else {
                console.log("Некорректный ввод. Пожалуйста, введите пару в формате 'Начальный пункт,Конечный пункт'.");
                i--; // Repeat input for this pair
            }
        }

        const result = findLongestChainDestination(itinerary);

        if (result === "обратитесь к специалисту") {
            console.log(result);
        } else if (!result) {
            console.log("Маршрут не найден.");
        } else if (itinerary.length === 1 && itinerary[0][0] === itinerary[0][1]) {
            console.log("обратитесь к специалисту");
        }
        else {
            console.log(`Итоговый конечный пункт: ${result}`);
        }
    } finally {
        rl.close();
    }
}

main();
